src/codec.rs

```diff
@@ -331,6 +335,135 @@ pub trait Decode: Sized {
 	}
 }
 
+/// Decode some data while having a generic context.
+pub trait DecodeWithContext<C>: Sized {
+	/// The type that will be decoded.
+	///
+	/// Attempt to deserialise the value from input while having a context.
+	fn decode_with_context<I: Input>(
+		input: &mut I,
+		ctx: &mut C,
+	) -> Result<Self, Error>;
+}
+
+#[impl_trait_for_tuples::impl_for_tuples(18)]
+impl<C> DecodeWithContext<C> for Tuple {
+	for_tuples!(where #( Tuple: DecodeWithContext<C> )*);
+	fn decode_with_context<I: Input>(
+		input: &mut I,
+		ctx: &mut C,
+	) -> Result<Self, Error> {
+		Ok((for_tuples!( #( Tuple::decode_with_context(input, ctx)? ),* )))
+	}
+}
+
+/// Decode some data while being constrained on the allocated memory.
+pub trait DecodeMemLimited: Sized {
+	const TYPE_INFO: TypeInfo = TypeInfo::Unknown;
```

Comment:
It looks a bit odd to me that we need this here, but without it, the `decode_vec_with...` function calls goes into the `decode_unoptimized` branch instead of doing the unsafe cast.